internal/pubsub/gcppubsub/gcppubsub.go

```diff
@@ -175,18 +178,39 @@ func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
 	for _, id := range ids {
 		ids2 = append(ids2, id.(string))
 	}
-	req := &pb.AcknowledgeRequest{
+	return s.client.Acknowledge(ctx, &pb.AcknowledgeRequest{
 		Subscription: s.path,
 		AckIds:       ids2,
-	}
-	err := s.client.Acknowledge(ctx, req)
-	if err != nil {
-		return fmt.Errorf("gcppubsub: making RPC to acknowledge messages: %v", err)
-	}
-	return nil
+	})
 }
 
 // Close implements driver.Subscription.Close.
 func (s *subscription) Close() error {
 	return s.client.Close()
 }
+
+// IsRetryable implements driver.Topic.IsRetryable.
+func (t *topic) IsRetryable(err error) (bool, time.Duration) {
+	return isRetryable(err), 0
+}
+
+// IsRetryable implements driver.Subscription.IsRetryable.
+func (s *subscription) IsRetryable(err error) (bool, time.Duration) {
+	return isRetryable(err), 0
+}
+
+// Copied from cloud.google.com/go/pubsub.
+func isRetryable(err error) bool {
+	s, ok := status.FromError(err)
+	if !ok {
+		return true
+	}
+	switch s.Code() {
+	case codes.DeadlineExceeded, codes.Internal, codes.ResourceExhausted:
```

Comment:
How long is the list of codes? Since subscriptions/topics fail hard for isRetryable=false, I'm wondering if a blacklist instead of a whitelist is more appropriate. E.g., if they add a new error code, we should default to retrying.

Reply:
The default should not be to retry. Failing fast is safer, because RPCs are not idempotent in general. That is why everyone gives whitelists for retryable codes. E.g. see the gRPC design I linked above.

Reply:
Interesting. Could/should there be a different default for `SendAck` + `ReceiveBatch` (which **are** idempotent, or roughly so) than for `SendBatch`, which is not?

Failing fast doesn't seem safer to me for an application that is expected to exit/crash when our API returns an error, and retry invisibly for retryable errors.

Reply:
Ugh. You reminded me that the GCP client does retries (and in fact uses different codes for different RPCs), so there's nothing to do. I removed this function.

But to continue the discussion: by failing fast being safer, I'm thinking of the case where you have a non-idempotent operation, like incrementing something. If there is an error code that leaves it ambiguous whether the operation actually happened, it would be better not to retry, rather than to do something wrong (double-count).

Reply:
> what else could they do?

Alert a human. Automatically retrying non-idempotent operations introduces silent corruption into the system.

Reply:
Isn't `ReceiveBatch` idempotent?

Reply:
I believe it is not. For example if there is one message in the subscription, and you call `ReceiveBatch` then it will pull that message down and it will be marked on the server as being processed. Now say another message shows up on the server and you call `ReceiveBatch` again. Then the new message gets marked as being processed. So that's not idempotent. It's time-dependent.